cpp/inline/games/stochastic_nim/PerfectPlayer.inl

```diff
@@ -0,0 +1,66 @@
+#include <games/stochastic_nim/PerfectPlayer.hpp>
+
+namespace stochastic_nim {
+
+inline PerfectPlayer::ActionResponse PerfectPlayer::get_action_response(const State& state,
+                                                                 const ActionMask& valid_actions) {
+  if (state.current_mode == kChanceMode) {
+    throw std::invalid_argument("PerfectPlayer does not support chance mode");
+  }
+
+  if (state.stones_left > strategy_->get_params().starting_stones) {
+    throw std::invalid_argument("PerfectPlayer does not support more stones than starting stones");
+  }
+
+  return strategy_->get_optimal_action()[state.stones_left] - 1;
+}
+
+PerfectStrategy::PerfectStrategy(Params params) : params_(params) {
+  state_value_ = new float[params_.starting_stones + 1];
+  optimal_action_ = new int[params_.starting_stones + 1];
```

Comment:
Whenever you have `new`, you need a corresponding `delete`. The right place for that would be in the destructor.

However, if you directly use the `Constants.hpp` values like I suggest, then you don't need these lines, and so you don't need a destructor.